common/cache/namespaceCache.go

```diff
@@ -492,16 +501,34 @@ UpdateLoop:
 	c.cacheByID.Store(newCacheByID)
 	c.cacheNameToID.Store(newCacheNameToID)
 	c.triggerNamespaceChangeCallbackLocked(prevEntries, nextEntries)
+
+	// only update last refresh time when refresh succeeded
+	c.lastRefreshTime.Store(now)
 	return nil
 }
 
 func (c *namespaceCache) checkNamespaceExists(
 	name string,
 	id string,
-) error {
+) (bool, error) {
```

Comment:
I think logic of this func should be changed. It is great that it returns `(bool, error)` now but `bool` should mean exist/not exist and error should indicate the error: persistence or throttling. I would return `service.ResourceExhausted` right from here and propagate it up to the caller. Caller should be able to differentiate if it is actual `NotFound` or "you are pinging me to fast".

Reply:
> but bool should mean exist/not exist

not really, bool means continue to check the cache & refresh the cache or not

Reply:
and there is not guarantee the `check DB see if the namespace exists` logic will be exercised within this function

Reply:
ok, then don't add this `bool` at all. Let it return `serviceerror.NotFound` or `serviceerror.ResourceExhausted` which will be propagated to the caller.

Reply:
nop, that is not the purpose of this function

the purpose of this function is check whether to continue the path, i.e. check & potentially force refresh the cache.
the bool return value means whether to continue (either the domain exists, or somebody has already checked within short period of time and it is really likely cache can contain the target domain, but no guarantee)

Reply:
the functionality (function in question) was originally introduced to solve the issue of 
t = 0, cache got refreshed, to be refreshed again at t = 10
t = 1, new domain got registered
t = 2, caller just want to use the domain (from within cache)

the function in question (along with the following logic) will solve the above case

Reply:
may be the function name should be changed?

Reply:
Yes, change the name. If `existsSomething()` returns `bool` it usually means exists/doesn't exist.

Also you return `NotFound` at t=2 now, right? I think `Unavailable` suites better here (client should wait and try again w/o performing extra actions): https://pkg.go.dev/google.golang.org/grpc/codes (comment to `FailedPrecondition Code = 8`).

Reply:
function name is changed

Unavailable means server fault, while in this case (namespace missing in cache) is client fault.
the only guarantee we have with customer is: if you namespace is in the DB, it is guaranteed to be in the cache after 10s + creation time

Reply:
I think clients should not know about cache or any internal details. So if the statement is "namespace creation can take up to 10 seconds" then returning `NotFound` within those 10 seconds makes perfect sense.